### ISSUE TITLE
Getting information on what triggers a given extension or other

### ISSUE THREAD

**Issue by cybik** [LukasBanana/XShaderCompiler#20]
Is there an option/way for this shader compiler to output reasons for a given extension or another ( GL_ARB_shading_language_420pack for example ) being pulled in?

**Comment by LukasBanana** [comment]
This is currently not part of the code reflection, if you mean that.

What should such a reflection look like, when you say 'outputing a *reason*'?
I could think of it like this:
```cpp
enum ExtensionRequirementReasonFlags {
    EarlyDepthStencil      = (1 << 0),
    MultiDimensionalArrays = (1 << 1),
    ExplicitBinding        = (1 << 2),
    BitwiseOperators       = (1 << 3),
    InitializerList        = (1 << 4),
    /* ... */
};

std::map<std::string, Flags> extensionRequirementReasonMap;

/* ... */

// Add 'initializer list' as reason for the
// "GL_ARB_shading_language_420pack" extension to the flags.
extensionRequirementReasonMap["GL_ARB_shading_language_420pack"] << InitializerList;
```
Is that what you are asking for?

If you just want to know which extensions or rather GLSL versions are choosen by the compiler for specific language features, take a look at [GLSLExtensionAgent.cpp](https://github.com/LukasBanana/XShaderCompiler/blob/master/src/Compiler/Backend/GLSL/GLSLExtensionAgent.cpp).

**Comment by cybik** [comment]
What I mean is, could there be some output, either in the console or in the generated shaders, that could explain why a given extension was introduced? It would be helpful to know so that some HLSL shaders can get altered to "fit" an earlier standard of GLSL.

Example of use case: We've got HLSL shaders right now, that by and large get converted to GLSL without any issues, but we have like 8 that need to bring 420 and I'm wondering exactly why that's pulled in, and more specifically which parts of the HLSL shader force 420 to be used.

**Comment by LukasBanana** [comment]
Ok, as I said there is currently no output from the compiler that tells you what you want to know.

But I can tell you that there are currently only two reasons for the extension `GL_ARB_shading_language_420pack` you mentioned:
1. Explicit binding slots (`-EB` shell option)
2. Initializer lists, e.g. `float2 x = { 1, 2 }` instead of `float2 x = float2(1, 2)`

I think your problem arises from the latter one, because GLSL only supports initializer lists since GLSL 4.20,
but I'm already planning to avoid these initializers.
However, this requires a little more conversion, but I'm working on this ;-)

**Comment by cybik** [comment]
Baller, thanks for the intel.

The suggestion stands though, but I'll try changing the source shaders to reflect this.

**Comment by LukasBanana** [comment]
Yes, I think it's a good idea to add this to the code reflection, but only when a specific shader output version is set (e.g. with the `-Vout` shell option) and extensions are enabled.

**Comment by LukasBanana** [comment]
The compiler in the master branch now supports translation of initializer lists, to avoid this extension.
However, translation of initializer lists is still restricted, i.e. nested initializers are currently not translated correcty (e.g. `{ 1, { 1, 2 } }` will result in a compiler error).

But maybe I will add your suggestion to the error output, when extensions are disabled, but the version is set too low.
This might look like this in the compiler output:
* `code generation error : GLSL extension 'GL_EXT_gpu_shader4' or shader output version 'GLSL 1.30' required (required for bitwise operator)`
* `code generation error : GLSL extension 'GL_ARB_shading_language_420pack' or shader output version 'GLSL 4.20' required (required for explicit binding slot)`

**Comment by cybik** [comment]
That would be awesome and grand.

**Comment by LukasBanana** [comment]
Solved with [4a8e440](https://github.com/LukasBanana/XShaderCompiler/commit/4a8e440348177d2c0db15f412e792e903a33433c).

Here is an example:

![ExtensionErrorExample](http://i.imgur.com/cJ5RA5y.png)

**Comment by LukasBanana** [comment]
With commit [ead2cbb](https://github.com/LukasBanana/XShaderCompiler/commit/ead2cbbb4aad4450fca8a9bdd93664b7c2804e95) there are now several warning options.
One is your requested extension feedback, which can be enabled with `-Wall` or `-Wextension`.
Use `xsc --help` for more details.
